### PR TITLE
Benchmark tasks for Matching TradeExecutor and OrderProcessor 

### LINKED ISSUE #2134: Benchmark trading daemons

### PR DESCRIPTION
**Pull request by ysv**
Implement Matching bench
* Implement TradeExecutor bench
* Implement OrderProcessor bench
* Implement Dummy & Bitfinex Order injecotrs
* Add bench config templates
* Add bitfinex injector data
* Expose RabbitMQ management API admin panel
* Implement Deposit & Member factories
* Add simple RabbitMQHTTP client
* Fix failures:trade_errors rake task
* Temporary skip feature specs

Co-authored-by: shal <ashanaakh@heliostech.fr>
Co-authored-by: ysv <ysavchuk@heliostech.fr>
Co-authored-by: mnaichuk <mnaichuk@heliostech.fr>
closes #2101 
closes #2134

### CODE REVIEW

File: config/templates/application.yml.erb [hunk 1]
```diff
@@ -171,12 +171,14 @@ defaults: &defaults
...
 
   # Event API configuration.
+  # TODO: Move Event API config to separate file.
 
   # JWT configuration.
   # You can generate keypair using:
   #
   #   ruby -e "require 'openssl'; require 'base64'; OpenSSL::PKey::RSA.generate(2048).tap { |p| puts '', 'PRIVATE RSA KEY (URL-safe Base64 encoded, PEM):', '', Base64.urlsafe_encode64(p.to_pem), '', 'PUBLIC RSA KEY (URL-safe Base64 encoded, PEM):', '', Base64.urlsafe_encode64(p.public_key.to_pem) }"
   #
+  EVENT_API_ENABLED: false
```
Comment (shal) [inline config/templates/application.yml.erb]: Could you make it as string `'false'`, otherwise this will show warnings after rails 5 patch.

Comment (shal) [inline config/templates/application.yml.erb]: Could you make it as string `'false'`, otherwise this will show warnings after rails 5 patch.

Comment (ysv) [reply config/templates/application.yml.erb]: @shal I would like to move Event API config into separate file

File: config/bench/data/usd-eur.csv [hunk 2]
```diff
@@ -0,0 +1 @@
...624, -106.37904833], [22964942193, 8.5623, -117.93918179], [22965617507, 8.5339, 100], [22965617569, 8.5339, 100], [22965617487, 8.5628, -106.37904833], [22965616799, 8.5749, -34.77130565], [22965617645, 8.5339, 100], [22965617678, 8.5692, -17.38579125], [22965617697, 8.5629, -106.37904833], [22965617704, 8.5339, 100], [22965617723, 8.5339, 100], [22965617749, 8.5339, 100], [22965617785, 8.5339, 100], [22965617697, 8.5628, -106.37904833], [22965617835, 8.5335, 100], [22965617817, 8.5458, -6.98943546], [22965617869, 8.5335, 100], [22965617697, 8.5626, -106.37904833], [22965617912, 8.5335, 100], [22965617924, 8.5335, 90], [22965617962, 8.5335, 100], [22965617973, 8.5335, 100], [22965617697, 8.5624, -106.37904833], [22965618058, 8.5335, 100], [22965618107, 8.5336, 16.42], [22965618119, 8.5336, 90], [22965617697, 8.5622, -106.37904833], [22965618177, 8.5335, 100], [22964942193, 8.5621, -117.93918179], [22965618214, 8.5458, -6.98795935], [22965618230, 8.5335, 100], [22965617697, 8.562, -106.37904833], [22965618314, 8.5684, -17.38728633], [22965617697, 8.5618, -106.37904833], [22965618375, 8.5335, 100], [22964942193, 8.5617, -117.93918179], [22965618420, 8.5336, 16.42], [22965618433, 8.5
```
Comment (ysv) [inline config/bench/data/usd-eur.csv]: @mnaichuk something is wrong with this data
1. in trade_executor bench nothing matched for 800 orders
2. for matching there was order with 0 price or (at leas I saw error: Cannt lock funds 0.0)

File: config/bench/matching.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  orders:
+    #
+    # Generic config.
+    injector:       bitfinex  # Order injector name.
+    number:         10000     # Number of orders to be created and published to RabbitMQ (at least 100_000).
+    markets:        usdeur    # Markets for performing trading.
+    #
+    # Bitfinex injector specific config.
+    # Path to bitfinex order history file.
+    # TODO: Add example how to generate this file.
```
Comment (ysv) [inline config/bench/matching.yml]: @mnaichuk

File: config/bench/order_processing.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  orders:
+    #
+    # Generic config.
+    injector:       bitfinex  # Order injector name.
+    number:         1200      # Number of orders to be created and published to RabbitMQ (at least 2000).
+    markets:        usdeur    # Markets for performing trading.
+    #
+    # Bitfinex injector specific config.
+    # Path to bitfinex order history file.
+    # TODO: Add example how to generate this file.
```
Comment (ysv) [inline config/bench/order_processing.yml]: @mnaichuk

File: config/bench/trade_execution.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  orders:
+    #
+    # Generic config.
+    injector:       bitfinex  # Order injector name.
+    number:         800       # Number of orders to be created and published to RabbitMQ (at least 1000).
+    markets:        usdeur    # Markets for performing trading.
+    #
+    # Bitfinex injector specific config.
+    # Path to bitfinex order history file.
+    # TODO: Add example how to generate this file.
```
Comment (ysv) [inline config/bench/trade_execution.yml]: @mnaichuk

File: lib/bench/factories.rb [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+      def construct_deposit
+        { amount: DEFAULT_DEPOSIT_AMOUNT,
+          type: 'Deposits::Fiat' }.merge(@options)
+      end
+    end
+  end
+end
+
+# Table name: deposits
```
Comment (ysv) [inline lib/bench/factories.rb]: @mnaichuk remove this

File: lib/bench/injectors.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
+module Bench
+  module Injectors
+    autoload :Dummy, 'bench/injectors/dummy'
+
+    class << self
+      # TODO: Rename.
+      def initialize_injector(config)
```
Comment (ysv) [inline lib/bench/injectors.rb]: @mnaichuk try to find better name or remove TODO

File: lib/bench/injectors/bitfinex.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
+require_relative '../injectors'
+
+module Bench
+  module Injectors
+    class Bitfinex < Base
+
+      def initialize(config)
+        super
+        if config[:data_load_path].present?
+          @data = YAML.load_file(Rails.root.join(config[:data_load_path]))
```
Comment (ysv) [inline lib/bench/injectors/bitfinex.rb:10]: @mnaichuk raise Bench::Error in case file doesn't exist

File: lib/bench/injectors/dummy.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
+require_relative '../injectors'
```
Comment (ysv) [inline lib/bench/injectors/dummy.rb]: @mnaichuk remove

File: lib/bench/matching.rb [hunk 10]
```diff
@@ -0,0 +1,117 @@
+require 'rabbitmq/http/client'
+
+# TODO: Add Bench::Error and better errors processing.
```
Comment (ysv) [inline lib/bench/matching.rb:3]: @mnaichuk

File: lib/bench/matching.rb [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+# TODO: Add Bench::Report and extract all metrics to it.
```
Comment (ysv) [inline lib/bench/matching.rb:4]: @mnaichuk

File: lib/bench/matching.rb [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+# TODO: Add missing frozen_string literal to whole module.
```
Comment (ysv) [inline lib/bench/matching.rb:5]: @mnaichuk

File: lib/bench/matching.rb [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+module Bench
+  class Matching
+    def initialize(config)
+      @config = config
+
+      # TODO: Use Faraday instead of RabbitMQ::HTTP::Client.
```
Comment (ysv) [inline lib/bench/matching.rb]: @shal

File: lib/bench/matching.rb [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+      @rmq_http_client = ::URI::HTTP.build(
+        scheme:   :http,
+        host:     ENV.fetch('RABBITMQ_HOST', 'localhost'),
+        port:     15672,
+        userinfo: "#{ENV.fetch('RABBITMQ_USER', 'guest')}:#{ENV.fetch('RABBITMQ_PASSWORD', 'guest')}"
+      ).yield_self { |endpoint| RabbitMQ::HTTP::Client.new(endpoint.to_s) }
+
+      @injector = Injectors.initialize_injector(@config[:orders])
+      @currencies = Currency.where(id: @config[:currencies].split(',').map(&:squish).reject(&:blank?))
+      # TODO: Print errors in the end of benchmark and include them into report.
```
Comment (ysv) [inline lib/bench/matching.rb:15]: @shal

File: lib/bench/matching.rb [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+      @errors = []
+    end
+
+    def run!
+      # TODO: Check if Matching daemon is running before start (use queue_info[:consumers]).
```
Comment (ysv) [inline lib/bench/matching.rb:20]: @shal

File: lib/bench/matching.rb [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+      Kernel.puts "Creating members ..."
+      @members = Factories.create_list(:member, @config[:traders])
+
+      Kernel.puts "Depositing funds ..."
+      @members.map(&method(:become_billionaire))
+
+      Kernel.puts "Generating orders by injector and saving them in db..."
+      # TODO: Add orders generation progress bar.
```
Comment (ysv) [inline lib/bench/matching.rb:28]: @shal

File: lib/bench/matching.rb [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+      @injector.generate!(@members)
+
+      @orders_number = @injector.size
+
+      Kernel.puts "Publishing messages to RabbitMQ..."
+      @publish_started_at = Time.now
+      # TODO: Add orders publishing progress bar.
```
Comment (ysv) [inline lib/bench/matching.rb:35]: @shal

File: lib/bench/matching.rb [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+            Kernel.puts e
+            @errors << e
+          end
+        end
+      end.map(&:join)
+    end
+
+    # TODO: Find better solution for getting message number in queue.
+    # E.g there is rabbitmqctl list_queues.
+    # TODO: Write useful queue info stats into file.
```
Comment (ysv) [inline lib/bench/matching.rb]: @shal you need to log queu stats once in 5sec for example into some file in log/bench

File: lib/bench/matching.rb [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    end
+
+    # TODO: Add more useful metrics to result.
+    def result
+      @result ||=
+      begin
+        publish_ops =  @orders_number / (@publish_finished_at - @publish_started_at)
+        matching_ops =  @orders_number / (@matching_finished_at - @publish_started_at)
+
+        # TODO: Deal with calling iso8601(6) everywhere.
```
Comment (ysv) [inline lib/bench/matching.rb:89]: @shal

File: lib/bench/matching.rb [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    end
+
+    # TODO: Implement method for saving report to file.
+    # File path should be configured from config.
+    def save_report
+    end
+
+    private
+    # TODO: Use get queue by name.
+    # TODO: Use Faraday instead of RabbitMQ::HTTP::Client.
```
Comment (ysv) [inline lib/bench/matching.rb:119]: @shal

File: lib/bench/matching.rb [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+          publish_started_at: @publish_started_at.iso8601(6),
+          publish_finished_at: @publish_finished_at.iso8601(6),
+          matching_started_at: @publish_started_at.iso8601(6),
+          matching_finished_at: @matching_finished_at.iso8601(6),
+          publish_ops: publish_ops,
+          matching_ops: matching_ops }
+      end
+    end
+
+    # TODO: Implement method for saving report to file.
```
Comment (ysv) [inline lib/bench/matching.rb]: @mnaichuk

File: lib/bench/order_processing.rb [hunk 22]
```diff
@@ -0,0 +1,76 @@
+module Bench
+  class OrderProcessing < TradeExecution
+    def run!
+      # TODO: Check if OrderProcessing daemon is running before start (use queue_info[:consumers]).
```
Comment (ysv) [inline lib/bench/order_processing.rb:6]: @shal

File: lib/bench/trade_execution.rb [hunk 23]
```diff
@@ -0,0 +1,45 @@
+module Bench
+  class TradeExecution < Matching
+    def run!
+      # TODO: Check if TradeExecutor daemon is running before start (use queue_info[:consumers]).
```
Comment (ysv) [inline lib/bench/trade_execution.rb:6]: @shal

File: lib/tasks/bench.rake [hunk 24]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+# TODO: Add descriptions.
```
Comment (ysv) [inline lib/tasks/bench.rake:3]: @shal

File: lib/tasks/bench.rake [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+# TODO: Save reports directly after running bench (in case next will error).
```
Comment (ysv) [inline lib/tasks/bench.rake]: @mnaichuk

File: lib/bench/order_processing.rb (hunk 22 above)
Comment (shal) [reply lib/bench/order_processing.rb:6]: Done

File: lib/bench/trade_execution.rb (hunk 23 above)
Comment (shal) [reply lib/bench/trade_execution.rb:6]: Done

File: lib/tasks/bench.rake (hunk 24 above)
Comment (shal) [reply lib/tasks/bench.rake:3]: Done.

File: lib/bench/matching.rb (hunk 15 above)
Comment (shal) [reply lib/bench/matching.rb:20]: Done

File: lib/bench/matching.rb (hunk 16 above)
Comment (shal) [reply lib/bench/matching.rb:28]: Done.

File: lib/bench/matching.rb (hunk 17 above)
Comment (shal) [reply lib/bench/matching.rb:35]: Done.

File: lib/bench/matching.rb (hunk 18 above)
Comment (shal) [reply lib/bench/matching.rb]: Done.

File: lib/bench/matching.rb (hunk 20 above)
Comment (shal) [reply lib/bench/matching.rb:119]: Done.

File: lib/bench/matching.rb (hunk 13 above)
Comment (shal) [reply lib/bench/matching.rb]: Done.

File: lib/bench/matching.rb (hunk 19 above)
Comment (shal) [reply lib/bench/matching.rb:89]: I don't understand how we can deal with it globally.

**Comment by ysv** [comment]
Since 
https://github.com/rubykube/peatio/pull/2106/commits/dd7f763e876d6279b0075125245ac624dccc1a5e
report looks like:
```yaml
---
config:
  currencies: usd,eur
  traders: 25
  threads: 50
  report_path: reports
  orders:
    injector: dummy
    number: 1200
    markets: usdeur
    min_volume: 0.1
    max_volume: 1.0
    min_price: 0.5
    max_price: 2.0
submit_publish:
  started_at: '2019-03-12T17:14:03.786892+02:00'
  finished_at: '2019-03-12T17:14:03.786892+02:00'
  operations: 1200
  ops: 5247.645119252736
matching:
  started_at: '2019-03-12T17:14:03.786892+02:00'
  finished_at: '2019-03-12T17:14:08.649498+02:00'
  operations: 1200
  ops: 246.78125268631675
trade_execution:
  started_at: '2019-03-12T17:14:03.786892+02:00'
  finished_at: '2019-03-12T17:14:43.508181+02:00'
  operations: 986
  ops: 24.822960805728133
cancel_publish:
  started_at: '2019-03-12T17:14:43.523195+02:00'
  finished_at: '2019-03-12T17:14:43.584340+02:00'
  operations: 212
  ops: 3467.168206721727
order_processing:
  started_at: '2019-03-12T17:14:43.523195+02:00'
  finished_at: '2019-03-12T17:14:53.866818+02:00'
  operations: 212
  ops: 20.49571992328027
```

**Pull request merged**